File: ispitch-api/src/analysis/application/rest/schemas/analysis.py
from datetime import datetime
from typing import Optional

from .....core.schemas import CamelCaseModel
from ..schemas.fillerwords import FillerWordsAnalysisSchema
from ..schemas.lexical_richness import LexicalRichnessAnalysisSchema
from ..schemas.silence import SilenceAnalysisSchema
from ..schemas.topic import TopicAnalysisSchema
from ..schemas.vocabulary import VocabularyAnalysisSchema
from ..schemas.prosody import ProsodyAnalysisSchema


class SpeechAnalysisSchema(CamelCaseModel):
    silence_analysis: Optional[SilenceAnalysisSchema] = None
    fillerwords_analysis: Optional[FillerWordsAnalysisSchema] = None
    vocabulary_analysis: Optional[VocabularyAnalysisSchema] = None
    lexical_richness_analysis: Optional[LexicalRichnessAnalysisSchema] = None
    topic_analysis: Optional[TopicAnalysisSchema] = None


class AudioAnalysisSchema(CamelCaseModel):
    speech_rate: float
    duration: float
    prosody_analysis: Optional[ProsodyAnalysisSchema] = None


class AnalysisSchema(CamelCaseModel):
    id: str
    user_id: Optional[str] = None
    status: str
    filename: str
    created_at: datetime
    updated_at: datetime
    transcription: Optional[str] = None
    speech_analysis: Optional[SpeechAnalysisSchema] = None
    audio_analysis: Optional[AudioAnalysisSchema] = None

File: frontend/src/domain/analysis/types/analysis.ts
interface Silence {
    start: number;
    end: number;
    duration: number;
}

export interface SilenceAnalysis {
    duration: number;
    silences: Silence[];
    pauses: number;
}

export interface FillerwordOccurrence {
    start: number;
    end: number;
    word: string;
}

export interface FillerwordAnalysis {
    total: number;
    distribution: Record<string, number>;
    occurrences: FillerwordOccurrence[];
}

export interface VocabularySuggestion {
    word: string;
    count: number;
    alternatives: string[];
}

export interface VocabularyAnalysis {
    suggestions: VocabularySuggestion[];
}

export interface LexicalRichnessAnalysis {
    typeTokenRatio: number;
    uniqueWords: number;
    totalWords: number;
}

export interface Topic {
    topic: string;
    summary: string;
}

export interface TopicAnalysis {
    topics: Topic[];
}

export interface SpeechAnalysis {
    silenceAnalysis: SilenceAnalysis;
    fillerwordsAnalysis: FillerwordAnalysis;
    vocabularyAnalysis: VocabularyAnalysis;
    lexicalRichnessAnalysis: LexicalRichnessAnalysis;
    topicAnalysis: TopicAnalysis;
}

export interface ContourPoint {
  time: number;
  value: number;
}

export interface PitchAnalysis {
  meanPitch: number;
  minPitch: number;
  maxPitch: number;
  stdevPitch: number;
  stdevPitchSemitones: number;
  pitchContour: Array<{ time: number; pitch: number }>;
}

export interface IntensityAnalysis {
  meanIntensity: number;
  minIntensity: number;
  maxIntensity: number;
  stdevIntensity: number;
  intensityContour: Array<{ time: number; volume: number }>;
}

export interface ProsodyAnalysis {
  pitchAnalysis?: PitchAnalysis;
  intensityAnalysis?: IntensityAnalysis;
  vocalQuality?: VocalQualityAnalysis;
}

export interface VocalQualityAnalysis {
  jitter: number;
  shimmer: number;
  hnr: number;
}

export interface AudioAnalysis {
    speechRate: number;
    duration: number;
    prosodyAnalysis?: ProsodyAnalysis;
}

export interface Analysis {
    id: string;
    status: 'pending' | 'completed' | 'failed';
    filename: string;
    createdAt: string;
    transcription: string;
    speechAnalysis: SpeechAnalysis;
    audioAnalysis: AudioAnalysis;
    score: number;
}
